src/rpc/blockchain.cpp

```diff
@@ -1203,6 +1202,89 @@ UniValue getchaintips(const UniValue& params, bool fHelp)
     return res;
 }
 
+UniValue z_gettreestate(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() != 1)
+        throw runtime_error(
+            "z_gettreestate \"hash|height\"\n"
+            "Return information about the given block's tree state.\n"
+            "\nArguments:\n"
+            "1. \"hash|height\"          (string, required) The block hash or height. Height can be negative where -1 is the last known valid block\n"
+            "\nResult:\n"
+            "[\n"
+            "  {\n"
+            "    \"hash\": \"xxxx\",         (string) block hash of the tip\n"
+            "    \"height\": xxxx,         (numeric) height of the chain tip\n"
+            "    \"sprout\": {\n"
+            "       \"finalCommitmentRoot\": \"hex\" (string)\n"
+            "       \"finalCommitmentState\": \"hex\" (string)\n"
+            "       \"skipHash\": \"hash\" (string)\n"
+            "     },\n"
+            "    \"sapling\": {\n"
+            "       \"finalCommitmentRoot\": \"hex\" (string)\n"
+            "       \"finalCommitmentState\": \"hex\" (string)\n"
+            "       \"skipHash\": \"hash\" (string)\n"
+            "     }\n"
```

Comment:
Missing the ending `}`, and includes a leading `[` that shouldn't be there (the output is a single object).

Also, looking at this output again, I think it might make sense to group the commitment tree components. The intention of `skipHash` is to enable clients to skip over blocks that don't have relevant information, but the way I originally described it, we'd skip back to the last block that had _any_ information. It's far more likely that at some point in the future we have `finalCommitmentRoot` available for any height but `finalCommitmentState` only available at certain heights, and it's the latter that clients actually care about fetching. So we could instead structure this as:
```suggestion
            "{\n"
            "  \"hash\": \"xxxx\",         (string) block hash of the tip\n"
            "  \"height\": xxxx,         (numeric) height of the chain tip\n"
            "  \"sprout\": {\n"
            "    \"commitments\": {\n"
            "      \"finalRoot\": \"hex\" (string)\n",
            "      \"finalState\": \"hex\" (string)\n"
            "    },\n"
            "    \"skipHash\": \"hash\" (string)\n"
            "  },\n"
            "  \"sapling\": {\n"
            "    \"commitments\": {\n"
            "     \"finalRoot\": \"hex\" (string)\n",
            "     \"finalState\": \"hex\" (string)\n"
            "    },\n"
            "    \"skipHash\": \"hash\" (string)\n"
            "  }\n"
            "}\n"
```
and then define `skipHash` as the block of the most recent prior ancestor that has _all_ information for _any_ of the siblings of `skipHash` (i.e. at present, complete information for the `commitments` section, but in future potentially also complete information for `nullifiers`). These might occur at different heights, which would mean a client potentially needs to skip back multiple times to get to the next complete `commitments` section.

Alternatively, we could move `skipHash` into the `commitments` section, define it as pointing to a block that has _all_ information about its siblings, and then add another `skipHash` to a future `nullifiers` section (enabling separate skiplist navigation). I'm not sure whether it's likely though that we might use different caching strategies for commitment trees and nullifier sets/trees within a single shielded protocol, so the utility of this is probably minimal.

Reply:
I suggest that the skip should be defined as follows: scan backwards and stop at the first block for which the amount of information we have increases.

Reply:
The part of my review here that was ambiguous at the time was the position of `skipHash`. The refactor to a `commitments` sub-object still needs to be done.